content/so/0009-process-management.md

```diff
@@ -330,6 +317,174 @@ Os processos são guardados numa _red-black tree_ ordenada por _vruntime_, que p
 
 É ainda possível definir prioridades estáticas superiores às dinâmicas (modo utilizador) em contexto _real-time_ ("_soft_", no sentido que não é 100% _real-time_). Para isto, são necessários privilégios de núcleo.
 
+:::warning[Informação por Rever]
+
+A informação desta secção ainda não foi revista e pode estar incorreta ou mal apresentada.
+
+:::
+
+## Operações asseguradas pelo Gestor de Processos
+
+**fork()**  
+A operação fork() reserva uma entrada na tabela `proc` (Unix), verifica se o utilizador não excedeu o número máximo de subprocessos e atribui um valor ao `pid` (normalmente um incremento de um inteiro mantido pelo núcleo).  
+De seguida, copia o contexto do processo pai: como a região de código é partilhada, apenas é incrementado o contador do número de utilizadores que acedem a essa região, as restantes regiões são copiadas.  
+Finalmente, é retornado o `pid` do novo processo ao processo pai, e zero ao filho (esses valores são colocados nas pilhas respetivas).
+
+![Criação de processos](./imgs/0009/process_creation.svg#dark=1)
+
+**exit()**  
+A operação `exit()` termina um processo, executando as funções registadas pelo `atexit`, libertando todos os recursos (ficheiro, diretoria corrente, regiões de memória).  
+De seguida actualiza o ficheiro que regista a utilização do processador, memória e I/O.  
+Finalmente, envia signal death of child (SIGCHILD) ao processo pai (que por omissão é ignorado) e mantem o filho no estado zombie, até que o pai o encontre (obtendo informação sobre a terminação do filho).
+
+**wait()**  
+O operação `wait()` procura por filhos zombie:
+
+- se não há filho zombie, o pai fica bloqueado;
+- se não há filhos, a funçao retorna imediatamente.
+  O pid do filho e estado do exit são returnados através do wait e a estrutura `proc` do filho é finalmente libertada.
+
+**exec()**  
+A funçao `exec()` executa um novo programa no âmbito de um processo já existente:
+
+- primeiro, verifica se o ficheiro existe e é executável;
+- copia argumentos da chamada exec da pilha do utilizador para o núcleo (pois o contexto utilizador irá ser destruído);
+- liberta as regiões de dados e pilha ocupadas pelo processo;
+- reserva novas regiões de memória;
+- carrega o ficheiro de código executável;
+- copia os argumentos da pilha do núcleo para a pilha do utilizador.
+
+O processo fica no estado executável e o contexto software do mesmo mantém-se inalterado.
+
+**signal()**  
+Se o processo tem rotina de tratamento associada ao signal, o núcleo regista no contexto do processo que o signal ocorreu.
+Antes do processo receber de novo execução, o despacho salta para a rotina de tratamento do signal.
+
+**pthread_mutex**  
+Fechar e abrir mutex's são chamadas de sistema. O núcleo mantém o estado de cada trinco, bem como uma lista de tarefas bloqueadas por esse trinco.
```

Comment:
not quite, a implementação do pthread_mutex tem de ser bloqueante, mas o núcleo não tem de necessariamente saber se o trinco está bloqueado ou não (em Linux isto não acontece: metes só tarefas bloqueadas à espera de um sinal de outra com a syscall `futex`).

Não me parece possível implementar espera bloqueante sem algum timpo de estrutura que guarda as tarefas bloqueadas no núcleo, mas mesmo assim diria para deixar isto intecionalmente *underspecified* e dizer apenas que a espera bloqueante é conseguida com ajuda do núcleo